lib/Promises/Deferred.pm

```diff
@@ -4,186 +4,207 @@ package Promises::Deferred;
 
 use strict;
 use warnings;
-
-use Scalar::Util qw[ blessed reftype ];
-use Carp qw[ confess ];
-
+use Ref::Util qw/is_blessed_ref is_coderef/;
+use Promises::Deferred::Default;
 use Promises::Promise;
 
-use constant IN_PROGRESS => 'in progress';
-use constant RESOLVED    => 'resolved';
-use constant REJECTED    => 'rejected';
+use constant {
+    IN_PROGRESS => 'in progress',
+    RESOLVED    => 'resolved',
+    REJECTED    => 'rejected',
+};
+my %statuses= (
+    0 => IN_PROGRESS,
+    1 => RESOLVED,
+    2 => REJECTED
+);
+
+my @pending_callbacks;
+my $notify_sub= Promises::Deferred::Default->get_notify_sub;
+sub _set_backend {
+    my ( $class, $arg ) = @_;
+    my $backend = $arg->[0] or return;
+    unless ( $backend =~ s/^\+// ) {
+        $backend = 'Promises::Deferred::' . $backend;
+    }
 
-sub new {
-    my $class = shift;
-    bless {
-        resolved => [],
-        rejected => [],
-        status   => IN_PROGRESS
-    } => $class;
+    eval "use $backend; 1;" or return;
```

Comment:
Ignore the "constructor" comment please, the diff was misleading, ... but the comment about Module::Runtime still applies.